clustered/server/src/main/java/org/ehcache/clustered/server/store/ServerStore.java

```diff
@@ -0,0 +1,115 @@
+ * Chain is list of {@link Link}s such that the same key has
+ * multiple mappings attached.
+ *
+ * Physical representation looks something like this :
+ *
+ *  hash -> |payLoad1| - |payLoad2| - |payLoad3|
+ *  hash -> |payLoad1| - |payLoad2|
+ *
+ *  This only allows append & replaceAtHead operations rather than
+ *  normal map operations.
+ *
+ */
+public interface ServerStore {
+
+  /**
+   * Returns the Chain associated with the provided hash.
+   * An empty Chain if no mapping exists.
+   *
+   * @param key hashcode of the key
+   * @return the {@link Chain} associated with the hash
+   */
+  Chain get(long key);
+
+  /**
+   * Appends the provided binary to Chain associated with key atomically.
+   * While appending, the payLoad is stored in {@link Link}.
+   *
+   * @param key to which the payLoad has to be appended
+   * @param payLoad to be appended
+   */
+  void append(long key, ByteBuffer payLoad);
+
+  /**
+   * The Chain associated with key, previous to append is returned.
+   * The operation atomically appends payLoad and returns the previously associated Chain
+   * with the key.
+   *  Following block of instructions are atomically performed.
+   *  {
+   *    Chain prevChain = get(key);
+   *    append(key, payLoad);
+   *    return prevChain;
+   *  }
+   *
+   * @param key to which the payLoad has to be appended
+   * @param payLoad to be appended
+   * @return the Chain associated with the key before payLoad was appended
+   */
+  Chain getAndAppend(long key, ByteBuffer payLoad);
+
+  /**
+   * Replaces the provided Chain with the equivalent Chain present at the head.
+   * This operation is not guaranteed to succeed.
+   * The replaceAtHead is successful iff the Chain associated with the key has
+   * a sub-sequence of links present as expected..
+   *
+   * If below mapping is present:
+   *
+   *    hash -> |payLoadA| - |payLoadB| - |payLoadC|
+   *
+   * And replaceAtHead(hash, |payLoadC'|, |payLoadA| - |payLoadB| - |payLoadC|) is invoked
+   * then this operation will succeed & the final mapping would be:
+   *
+   *    hash -> |payLoadC'|
+   *
+   * The same operation will also succeed if the mapping was modified by the time replace was invoked to
+   *
+   *    hash -> |payLoadA| - |payLoadB| - |payLoadC| - |payLoadD|
+   *
+   * Though the final mapping would be:
+   *
+   *    hash -> |payLoadC'| - |payLoadD|
+   *
+   * Failure case:
+   *
+   * But before replaceAtHead if it was modified to :
+   *
+   *    hash -> |payLoadC"| - |payLoadD|
+   *
+   * then replaceAtHead(hash, |payLoadC'|, |payLoadA| - |payLoadB| - |payLoadC|) will be ignored.
+   *
+   * Chain to be replaced, should also adhere to the rule of ascending order of Links.
+   *
+   * @param key for which Chain has to be replaced
+   * @param update the new Chain to be replaced
+   * @param expect of the original Chain which this operation intends to replaceAtHead
+   */
+  void replaceAtHead(long key, Chain update, Chain expect);
```

Comment:
This ordering is a bit unconventional.  More normal for this kind of API would be expected then update.